src/marshmallow/fields.py

```diff
@@ -1634,6 +1638,91 @@ def __init__(self, *args, **kwargs):
         self.validators = [validator] + original_validators
 
 
+class IP(String):
+    """A IP address field.
+
+    :param bool exploded: If `True`, serialize ipv6 address in long form, ie. with groups
+        consisting entirely of zeros included."""
+
+    default_error_messages = {"invalid_ip": "Not a valid IP address."}
+
+    def __init__(self, *args, exploded=False, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.exploded = exploded
+
+    def _validated(
+        self, value
+    ) -> typing.Optional[typing.Union[ipaddress.IPv4Address, ipaddress.IPv6Address]]:
+        """Format the value or raise a :exc:`ValidationError` if an error occurs."""
+        if value is None:
+            return None
+        try:
+            if isinstance(value, (int, bytes)):
+                # ip_address function is flexible in the terms of input value. In the case of
+                # marshalling, integer and binary address representation parsing may lead to
+                # confusion.
+                raise TypeError(
+                    "Only dot-decimal and hexadecimal groups notations are supported."
+                )
+            return ipaddress.ip_address(value)
+        except (ValueError, TypeError) as error:
+            raise self.make_error("invalid_ip") from error
+
+    def _serialize(self, value, attr, obj, **kwargs) -> typing.Optional[str]:
+        if value is not None and self.exploded:
+            value = value.exploded
+        return super()._serialize(value, attr, obj, **kwargs)
+
+    def _deserialize(
+        self, value, attr, data, **kwargs
+    ) -> typing.Optional[typing.Union[ipaddress.IPv4Address, ipaddress.IPv6Address]]:
+        return self._validated(value)
+
+
+class IPv4(IP):
+    """A IPv4 address field."""
+
+    default_error_messages = {"invalid_ip": "Not a valid IPv4 address."}
+
+    def _validated(self, value) -> typing.Optional[ipaddress.IPv4Address]:
+        """Format the value or raise a :exc:`ValidationError` if an error occurs."""
+        if value is None:
+            return None
+        try:
+            if isinstance(value, (int, bytes)):
+                raise TypeError("Only dot-decimal notation is supported.")
```

Comment:
What about hexadecimal groups as above in generic IP field?

Reply:
IPv6 error message mistake. Dot decimal notation is exclusive for IPv4 and hexadecimal groups are appropriate for IPv6. Fixed.